### PR TITLE
Fix gradient bug for distance_transform.

### PR DESCRIPTION
**Pull request by dichen-cd**
#### Changes
Fixed a gradient computation bug for `distance_transform`.

Fixes # (issue)

https://github.com/kornia/kornia/issues/1579

#### Type of change
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/contrib/distance_transform.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def distance_transform(
 
         offset: int = i * kernel_size // 2
         out += (offset + cdt) * mask
-        boundary[mask == 1] = 1
+        boundary = torch.where(mask == 1, torch.ones_like(boundary), boundary)
```
Comment (shijianjian) [inline kornia/contrib/distance_transform.py:71]: I would suggest to put torch.ones_like out of the for-loop to reduce the memory creation.

**Review by shijianjian — CHANGES_REQUESTED** [review]
Can you add a quick test for gradient check?

File: kornia/contrib/distance_transform.py (hunk 1 above)
Comment (dichen-cd) [reply kornia/contrib/distance_transform.py:71]: I'm not familiar with pytest or gradient check. 
Is it okay if I add a function under `kornia.test.test_contrib.TestConvDistanceTransform`?

```python
class TestConvDistanceTransform:
    def test_loss_grad(self, device, dtype):
        B, C, H, W = 1, 1, 32, 32
        input1 = torch.rand(B, C, H, W, device=device, dtype=dtype, requires_grad=True)
        input2 = torch.rand(B, C, H, W, device=device, dtype=dtype, requires_grad=True)
        tiny_module = nn.Conv2d(1, 1, (3, 3), (1, 1), (1, 1))
        input1 = kornia.contrib.distance_transform(tiny_module(input1))
        input2 = kornia.contrib.distance_transform(input2)
        loss = torch.nn.functional.mse_loss(input1, input2)
        try:
            loss.backward()
        except RuntimeError as e:
            print('Grad check failed for kornia.contrib.distance_transform when calculating regression loss.')
            raise e
```

Comment (shijianjian) [reply kornia/contrib/distance_transform.py:71]: LGTM :)

File: test/test_contrib.py [hunk 2]
```diff
@@ -436,6 +436,21 @@ def test_gradcheck(self, device, dtype):
...
 
+    def test_loss_grad(self, device, dtype):
+        B, C, H, W = 1, 1, 32, 32
+        input1 = torch.rand(B, C, H, W, device=device, dtype=dtype, requires_grad=True)
+        input2 = torch.rand(B, C, H, W, device=device, dtype=dtype, requires_grad=True)
+        tiny_module = nn.Conv2d(1, 1, (3, 3), (1, 1), (1, 1))
+        input1 = kornia.contrib.distance_transform(tiny_module(input1))
+        input2 = kornia.contrib.distance_transform(input2)
+        loss = torch.nn.functional.mse_loss(input1, input2)
+        try:
```
Comment (edgarriba) [inline test/test_contrib.py:447]:
```suggestion
        assert gradcheck(loss, (input1, input2,), raise_exception=True)
```
I believe this should work, or similar

File: kornia/contrib/distance_transform.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ def distance_transform(
 
     # It is possible to avoid cloning the image if boundary = image, but this would require modifying the image tensor.
     boundary = image.clone()
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:56]: We should evaluate wether this `.clone()` is needed

File: test/test_contrib.py (hunk 2 above)
Comment (edgarriba) [reply test/test_contrib.py:447]: And you might want to lower input sizes if you see it’s slow

Comment (dichen-cd) [reply test/test_contrib.py:447]: `gradcheck(mse_loss, (input1, input2,), raise_exception=True)` won't be able to find the bug during `loss.backward()`.
I think that's why this bug exists even when there's a grad check at https://github.com/kornia/kornia/blob/fc4b2fb7e7b0a8abbc802e67e8406a5a32c32d62/test/test_contrib.py#L434-L437

Comment (shijianjian) [reply test/test_contrib.py:447]: I see. In that case, gradcheck only check the gradients computation but not check if we can backward or not. So that there might be other non-backwardable functions in the lib.

Comment (edgarriba) [reply test/test_contrib.py:447]: Very interesting to see that since we rely quite a lot on `gradcheck` to verify differentiability. /cc @shijianjian @ducha-aiki @lferraz

Comment (edgarriba) [reply test/test_contrib.py:447]: I guess you are removing the `try/catch` block after the fix

File: kornia/contrib/distance_transform.py (hunk 3 above)
Comment (dichen-cd) [reply kornia/contrib/distance_transform.py:56]: I think it would be better to keep the `.clone()` here. 
If removed, the following loop would change the pixel values of image gradually, which IMO is not a desired behaviour.

File: test/test_contrib.py (hunk 2 above)
Comment (dichen-cd) [reply test/test_contrib.py]: The `try/catch` block is removed in https://github.com/kornia/kornia/pull/1584/commits/dcd1c5e17cf4d2ae2db1f438c53245bba0afd93f .
Up to now I'm not aware of a better way to check whether gradient could be back-propagated. 
Any suggestions?

**Review by edgarriba — APPROVED** [review]
LGTM. @shijianjian review and merge
